platform/keysystem/worldconfig/keyserver.go

```diff
@@ -0,0 +1,473 @@
+package worldconfig
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"net"
+	"os"
+
+	"github.com/sipb/homeworld/platform/keysystem/keyserver/account"
+	"github.com/sipb/homeworld/platform/keysystem/keyserver/authorities"
+	"github.com/sipb/homeworld/platform/keysystem/keyserver/config"
+	"github.com/sipb/homeworld/platform/keysystem/keyserver/verifier"
+	"github.com/sipb/homeworld/platform/keysystem/worldconfig/paths"
+)
+
+type Groups struct {
+	KerberosAccounts *account.Group
+	RootAdmins       *account.Group
+	Nodes            *account.Group
+	SupervisorNodes  *account.Group
+	WorkerNodes      *account.Group
+	MasterNodes      *account.Group
+}
+
+func GenerateGroups(context *config.Context) Groups {
+	kerberosAccountsGroup := &account.Group{
+		Name: "kerberos-accounts",
+	}
+	nodesGroup := &account.Group{
+		Name: "nodes",
+	}
+	groups := Groups{
+		KerberosAccounts: kerberosAccountsGroup,
+		Nodes:            nodesGroup,
+		RootAdmins: &account.Group{
+			Name:       "root-admins",
+			SubgroupOf: kerberosAccountsGroup,
+		},
+		SupervisorNodes: &account.Group{
+			Name:       "supervisor-nodes",
+			SubgroupOf: nodesGroup,
+		},
+		WorkerNodes: &account.Group{
+			Name:       "worker-nodes",
+			SubgroupOf: nodesGroup,
+		},
+		MasterNodes: &account.Group{
+			Name:       "master-nodes",
+			SubgroupOf: nodesGroup,
+		},
+	}
+
+	return groups
+}
+
+func GenerateAccounts(context *config.Context, conf *SpireSetup, groups Groups) error {
+	var accounts []*account.Account
+
+	// TODO: ensure that node hostnames are not duplicated
+
+	for _, node := range conf.Nodes {
+		var schedule string
+		var group *account.Group
+		if node.Kind == "worker" {
+			schedule = "true"
+			group = groups.WorkerNodes
+		} else {
+			schedule = "false"
+			if node.Kind == "supervisor" {
+				group = groups.SupervisorNodes
+			} else if node.Kind == "master" {
+				group = groups.MasterNodes
+			} else {
+				return fmt.Errorf("unrecognized kind of node: %s", node.Kind)
+			}
+		}
+
+		limitIP := net.ParseIP(node.IP)
+		if limitIP == nil {
+			return fmt.Errorf("invalid IP address: %s", node.IP)
+		}
+
+		principal := node.Hostname + "." + conf.Cluster.ExternalDomain
+
+		accounts = append(accounts, &account.Account{
+			Principal: principal,
+			Group:     group,
+			LimitIP:   limitIP,
```

Comment:
The semantics are that attempts to directly authenticate from any other IP address will be rejected. It's not the primary layer of security, but is a minor piece of defense in depth, and is primarily used to avoid accidentally configuring a server as the wrong server by entering a token in the wrong server.